src/main/java/com/box/sdk/BoxWebHookSignatureVerifier.java

```diff
@@ -0,0 +1,256 @@
+                webHookPayload, deliveryTimestamp)) {
+            return true;
+        }
+
+        // check secondary key signature if secondary key exists
+        if (this.secondarySignatureKey != null && this.verify(this.secondarySignatureKey, algorithm, secondarySignature,
+                webHookPayload, deliveryTimestamp)) {
+            return true;
+        }
+
+        // default strategy is false, to minimize security issues
+        return false;
+    }
+
+    /**
+     * Verifies a provided signature.
+     *
+     * @param key
+     *            for which signature key
+     * @param actualAlgorithm
+     *            current signature algorithm
+     * @param actualSignature
+     *            current signature
+     * @param webHookPayload
+     *            for signing
+     * @param deliveryTimestamp
+     *            for signing
+     * @return true if verification passed
+     */
+    private boolean verify(String key, BoxSignatureAlgorithm actualAlgorithm, String actualSignature,
+            String webHookPayload, String deliveryTimestamp) {
+        if (actualSignature == null) {
+            return false;
+        }
+
+        byte[] actual = Base64.decode(actualSignature);
+        byte[] expected = this.signRaw(actualAlgorithm, key, webHookPayload, deliveryTimestamp);
+
+        return Arrays.equals(expected, actual);
+    }
+
+    /**
+     * Calculates signature for a provided information.
+     *
+     * @param algorithm
+     *            for which algorithm
+     * @param key
+     *            used by signing
+     * @param webHookPayload
+     *            for singing
+     * @param deliveryTimestamp
+     *            for signing
+     * @return calculated signature
+     */
+    public String sign(BoxSignatureAlgorithm algorithm, String key, String webHookPayload, String deliveryTimestamp) {
+        return Base64.encode(this.signRaw(algorithm, key, webHookPayload, deliveryTimestamp));
+    }
+
+    /**
+     * Calculates signature for a provided information.
+     *
+     * @param algorithm
+     *            for which algorithm
+     * @param key
+     *            used by signing
+     * @param webHookPayload
+     *            for singing
+     * @param deliveryTimestamp
+     *            for signing
+     * @return calculated signature
+     */
+    private byte[] signRaw(BoxSignatureAlgorithm algorithm, String key, String webHookPayload,
+            String deliveryTimestamp) {
+        Mac mac = MAC_POOL.acquire(algorithm.javaProviderName);
+        try {
+            mac.init(new SecretKeySpec(key.getBytes(UTF_8), algorithm.javaProviderName));
+            mac.update(UTF_8.encode(webHookPayload));
+            mac.update(UTF_8.encode(deliveryTimestamp));
+        } catch (InvalidKeyException e) {
+            throw new IllegalArgumentException("Invalid key: ", e);
+        } finally {
+            MAC_POOL.release(mac);
+        }
+        return mac.doFinal();
```

Comment:
I guess mac.doFinal() should be performed before returning the Mac to the pool? Else it might lead to race condition, where some other thread re-uses this Mac (before we call mac.doFinal() in the current thread).